ppmclibs/tinycv_impl.cc

```diff
@@ -90,12 +90,7 @@ double enhancedMSE(const Mat& _I1, const Mat& _I2)
         const uchar* I2_data = I2.ptr<const uchar>(j);
 
         for (int i = 0; i < I1.cols; i++) {
-            // reduce the colours to 16 before checking the diff
-            if (abs(I1_data[i] - I2_data[i]) < 16)
-                continue; // += 0
-            double t1 = round(I1_data[i] / 16.);
-            double t2 = round(I2_data[i] / 16.);
-            double diff = (t1 - t2) * 16;
+            double diff = I1_data[i] - I2_data[i];
```

Comment:
This calculation was always a bit strange ...
- For small differences (abs < 16), the error is set to 0
- Otherwise, the inputs are quantized and then the difference is calculated. In some cases this reduces the error (when both are rounded towards the other, e.g. 39 and 9 -> 32-16=16), in some cases the exaggerates the error (41 and 7 -> 48-0=48).

AFAIK one/the motivation for the quantization was to suppress JPEG artifacts, introduced by some VNC/RFB connections with some BMCs.

Reply:
With the changed calculation, this is just a plain (normalized) sum-of-squared-differences, i.e. the exact same value which is returned by template_match.

I.e. you can scrap this function altogether and just use `double mse = sse / (width * height)` @ L260.

Reply:
I tried to simplify the code but ran into some overflow issues that I didn't want to continue debugging - so let's leave that to another PR.

Reply:
You can see here an example of a change that was ok before and is now marked as 92% (a bit unexpected to the human eye I would say):

<img src="https://stephan.kulow.org/ezgif-2-9482783f0c36.gif"/>

(animated gif with before and after frame)

Reply:
~49% of the pixels have a color difference of 0, ~49% of 1..15, and the remaining 1.7% have a difference of 16..19.

Algorithmically, the changed result is well expected.

Reply:
I'm not saying the algorithm is faulty - I'm trying to explain the motiviation of the old one :)

Reply:
Basically if we could get an easy math for perceived difference, we could win. I tend to believe the answer lies in the HSV model

Reply:
The "V" in "HSV" is more or less the same as the current gray channel.

For a distance metric, HSV is quite bad - e.g. dark yellow and dark blue have the maximum distance of 180 degress, although both are perceived as "almost black". The problem is not YUV, but taking only the lightness.

A good metric:
1. suppresses large error values due to small changes, e.g. for a large monochrome background being slightly darker/lighter
2. takes into account shifts in color. I.e. also us YU channels

The old code is fairly good at (1.), but fails for (2.). The proposed change no longer supports (1.), and does not help for (2.).

Reply:
It helps those cases where the shift in colour is light. But basically we're fighting a general colour blindless and so far we got pretty convincing if we defended's openQA colour blindness with it being a problem to (some) humans too.

If you upload the needle in question to https://www.color-blindness.com/coblis-color-blindness-simulator/ you'll get 
![image](https://user-images.githubusercontent.com/1067203/111039046-49695780-842c-11eb-8a5c-aa7a45bf3a76.png)

Now tell me: is the pattern selected or not?

Reply:
this form of colour blindness affects 1 out of 30.000 people, so it's not a strong argument. The more frequent Deuteranopia is simulated as
![image](https://user-images.githubusercontent.com/1067203/111039181-08be0e00-842d-11eb-9005-9b9cd5fad67a.png)

which leave more contrast in the checkbox.

Reply:
I have no vision impairments, and in my opinion the theming for the checkboxes is just horrible. Any impairments (color weakness/blindness, far/shotsighted, ...) just make it worse.

Reply:
I do not disagree.

My concern is just testers creating a needle come and say "how can this be 100% when it's clearly changing from grey to green" and so I prefer if we make such problems discoverable.

Reply:
@StefanBruens But are you fine if we'd merge it? The new behavior might not be a good metric for measuring the perceived difference but it helps with the use-case we currently see in production.